# How to Extract Low-Level Video Features using the EmoCodes Library
This notebook provides a step-by-step guide on how to automatically extract low-level video features from a video file. 

**Note**: This processing will take longer for larger video files.

In [1]:
import emocodes as ec
import pandas as pd

video_file = '/Users/catcamacho/Documents/GitHub/emocodes/testing/data/sample_clip1.mp4' # must be MP4
sampling_rate = 10 # in Hz

features = ec.ExtractVideoFeatures().extract_features(video_file, sampling_rate)

/Users/catcamacho/Library/Python/3.8/lib/python/site-packages/outdated/utils.py:14: OutdatedPackageWarning: The package pingouin is out of date. Your version is 0.4.0, the latest is 0.5.0.
Set the environment variable OUTDATED_IGNORE=1 to disable these warnings.
  return warn(


extracting video features...
Extracting brightness...


Stim: 1461it [00:16, 86.33it/s]


Extracting saliency...


Stim: 1461it [03:02,  8.01it/s]


Extracting sharpness...


Stim: 1461it [00:02, 539.10it/s]


Extracting vibrance...


Stim: 1461it [00:42, 34.14it/s]


Combining data...
Visual feature extraction complete.
done!
extracting audio features...
Extracting dynamic tempo...
Extracting loudness...
Extracting major music beats...
Aggregating data...
Auditory feature extraction complete.
done!


The above class produces a CSV with the video features called "/Users/catcamacho/Documents/GitHub/emocodes/testing/data/sample_clip1_features.csv". We can also access it from the class directly since we have it loaded up.

In [2]:
features.combined_df.head()

brightness  frac_high_saliency  sharpness    vibrance       rms  \
time_ms                                                                    
0.0        0.376637            0.352731        1.0  370.878880  0.064232   
100.0      0.372393            0.352521        1.0  363.655722  0.064523   
200.0      0.334205            0.340044        1.0  298.343229  0.067314   
300.0      0.297925            0.339049        1.0  240.070889  0.083041   
400.0      0.257110            0.331729        1.0  182.240036  0.071761   

         beats  dynamic_tempo  
time_ms                        
0.0        0.1     169.394531  
100.0      0.0     172.265625  
200.0      0.0     172.265625  
300.0      0.0     172.265625  
400.0      0.0     172.265625

We can now combine these features with the manual codes from this clip:

In [3]:
codes = pd.read_csv('/Users/catcamacho/Dropbox/Mac/Documents/GitHub/emocodes/testing/data/sample_codes_clip1.csv', index_col=0)
combined = codes.merge(features.combined_df, how='left', left_index=True, right_index=True)
combined.to_csv('/Users/catcamacho/Dropbox/Mac/Documents/GitHub/emocodes/testing/data/sample_codes_clip1_combinedfeatures.csv')

### Generate Report

In [4]:
in_file = '/Users/catcamacho/Dropbox/Mac/Documents/GitHub/emocodes/testing/data/sample_clip1_features.csv'
out_folder = '/Users/catcamacho/Dropbox/Mac/Documents/GitHub/emocodes/testing/data/full_report'

ec.SummarizeVideoFeatures().compile(in_file, out_folder)

<Figure size 504x324 with 0 Axes>

<Figure size 504x324 with 0 Axes>

Below is what the produced report looks like:

# EmoCodes Analysis Summary Report

**in_file:** /Users/catcamacho/Dropbox/Mac/Documents/GitHub/emocodes/testing/data/sample_clip1_features.csv 

| Feature | Non-Zero | Min Value | Max Value |
| :------ | :------: | :-------: | :-------: |
| brightness | 97.95% | 0.0 | 1.0 |
| frac_high_saliency | 97.95% | 0.0 | 0.4 |
| sharpness | 97.95% | 0.0 | 1.0 |
| vibrance | 95.67% | 0.0 | 3784.3 |
| rms | 100.0% | 0.0 | 0.2 |
| beats | 18.22% | 0.0 | 0.1 |
| dynamic_tempo | 100.0% | 112.3 | 172.3 |

******

## Features Included in this Analysis

### Original Features

![feature plots](llfigs/features_plot.png)

### After HRF convolution (6s peak, 12s undershoot)

![hrf-convolved feature plots](llfigs/hrf_features_plot.png)

******

## Spearman Correlations

![correlation plots](llfigs/corr_plot.png)

******
## Mean Instantaneous Phase Synchrony

![mean IPS plots](llfigs/mean_ips_plot.png)

******
## Variance Inflation Factors

![VIF plots](llfigs/vif_plot.png)
